### ISSUE TITLE
Storage: Client is not `Unpin`

### ISSUE THREAD

**Issue by thomastaylor312** [Azure/azure-sdk-for-rust#133]
Because we allow for streaming blobs back, we need to make sure the client can be unpinned. `stream.next()` requires `Self` to be `Unpin`, especially as these streams can be passed back to other clients (e.g. an http response back to a client fetching something from a service)

**Comment by rylev** [comment]
@thomastaylor312 can you provide a minimum viable reproduction case for this issue? Which client are you referring to? `azure_core::HttpClient`? `azure_cosmos::CosmosClient`?

**Comment by thomastaylor312** [comment]
I honestly don't know where it came from. Let me get something small for an example

**Comment by thomastaylor312** [comment]
Ok, here is a simple reproduction: https://github.com/thomastaylor312/blob-unpin. This is a stripped down version of what I was trying to do in real code. Namely, I was trying to get a stream that I could return as a response out of a function. When you attempt to compile, you'll note 2 main problems: First, the lifetime constraint from the `BlobStreamBuilder` is applied to all parameters and the returned stream, so the borrow doesn't live long enough to pass the stream back (so this is an additional bug, as I shouldn't need to specify a lifetime on my function to get a stream of data back). The second shows this original problem, something buried deep in the call chain (somewhere in the part of the code where we set up the stream) doesn't implement Unpin:

```
within `futures_util::stream::unfold::_::__Unfold<'_, Option<azure_core::prelude::Range>, [closure@BlobStreamBuilder<'<empty>, KeyClient, Yes, Yes, Yes>::finalize::{closure#1}], impl Future>`, the trait `Unpin` is not implemented for `from_generator::GenFuture<[static generator@BlobStreamBuilder<'<empty>, KeyClient, Yes, Yes, Yes>::finalize::{closure#1}::{closure#0} for<'r, 's, 't0> {ResumeTy, azure_core::prelude::Range, GetBlobBuilder<'r, KeyClient, Yes, Yes>, impl Future, ()}]>`
```

**Comment by MindFlavor** [comment]
Hi @thomastaylor312, sorry not having been able to look at this before. 
It's not that Client does not implement Unpin, it's the stream that doesn't. In order to consume it you need to pin it yourself with [`Box::pin`](https://doc.rust-lang.org/std/boxed/struct.Box.html?search=#method.pin).

Regarding your example, you can change line 14 from:

```rust
let mut stream = get_blob_stream(client);
```

to 


```rust
let mut stream = Box::pin(get_blob_stream(client));
```

This will let you consume the stream.

The slightly modified version of your example that compiles is like this:

```rust
use azure_core::errors::{AzureError, UnexpectedHTTPResult};
use azure_core::*;
use azure_storage::blob::prelude::*;
use azure_storage::client;
use azure_storage::key_client::KeyClient;
use futures::StreamExt;
use prelude::Range;
  
#[tokio::main]
async fn main() {
    let client = client::with_azure_sas("foo", "notatoken");

    let range = Range::new(0, 2000);
    let mut stream = Box::pin(get_blob_stream(&client, &range));
  
    while let Some(res) = stream.next().await {
        println!("{:?}", res.unwrap());
    }   
}
  
fn get_blob_stream<'a>(
    client: &'a KeyClient,
    range: &'a Range,
) -> impl futures::Stream<Item = Result<Vec<u8>, AzureError>> + 'a {
    let stream = client
        .stream_blob()
        .with_container_name("foo")
        .with_blob_name("foo")
        .with_range(range)
        .finalize();
    stream
}
```

(I had to constrain the stream to the lifetime of the borrows).

**Comment by thomastaylor312** [comment]
Hmmm...I didn't try pinning here. But when I did try to pin it before in my real code, it didn't seem to work for me (still complaining about `Unpin`). However, it did just work in my simple example. I'll be revisiting my real code here next week anyway and I can give it a try there again to see if this was just a stupid error on my part.

However, I think the usability of this with the lifetimes, particularly around ranges (which are likely to be created ad hoc) should be updated. Is that worth opening another issue?

**Comment by MindFlavor** [comment]
> However, I think the usability of this with the lifetimes, particularly around ranges (which are likely to be created ad hoc) should be updated. Is that worth opening another issue?

Yes I you are right, reusing ranges is unlikely. We should go with ownership. 

I'll fix it in the PR I still have pending (#119). Please note that PR will change the API substantially. For example there will be no more a specific "stream blob" function, you will just have to call `get().stream()` instead of `get().finalize()`.

**Comment by thomastaylor312** [comment]
Closing in favor of 164
